membership.md

```diff
@@ -0,0 +1,83 @@
+|   AddNonvoter        |         AddVoter                   \                 |
+|       |       ,->---' `--<-.      |                        \                |
+|       v      /              \     v                         \               |
+|  +----------+                +----------+                    +----------+   |
+|  |          | ---AddVoter--> |          | -log caught up --> |          |   |
+|  | nonvoter |                | staging  |                    |  voter   |   |
+|  |          | <-DemoteVoter- |          |                 ,- |          |   |
+|  +----------+         \      +----------+                /   +----------+   |
+|                        \                                /                   |
+|                         `--------------<---------------'                    |
+|                                                                             |
++-----------------------------------------------------------------------------+
+```
+
+While these operations aren't quite symmetric, we think they're a good set to capture
+the possible intent of the user. For example, if I want to make sure a server doesn't have a vote, but the server isn't part of the configuration at all, it probably shouldn't be added as a nonvoting server.
+
+Each of these application-level operations will be interpreted by the leader and, if it has an effect, will cause the leader to write a new configuration entry to its log. Which particular application-level operation caused the log entry to be written need not be part of the log entry.
+
+## Code implications
+
+This is a non-exhaustive list, but we came up with a few things:
+- Remove the PeerStore: the `peers.json` file introduces the possibility of getting out of sync with the log and snapshot, and it's hard to maintain this atomically as the log changes. It's not clear whether it's meant to track the committed or latest configuration, either.
+- Servers will have to search their snapshot and log to find the committed configuration and the latest configuration on startup.
+- Bootstrap will no longer use `peers.json` but should initialize the log or snapshot with an application-provided configuration entry.
+- Snapshots should store the index of their configuration along with the configuration itself. In my experience with LogCabin, the original log index of the configuration is very useful to include in debug log messages.
+- As noted in hashicorp/raft#84, configuration change requests should come in via a separate channel, and one may not proceed until the last has been committed.
+- As to deciding when a log is sufficiently caught up, implementing a sophisticated algorithm *is* something that can be done in a separate PR. An easy and decent placeholder is: once the staging server has reached 95% of the leader's commit index, promote it.
+
+## Feedback
+
+Again, we're looking for feedback here before we start working on this. Here are some questions to think about:
```

Comment:
> + - Does this seem like where we want things to go?

After a first look, I think so - there's a lot of utility with the extra states, and I think they provide nice ways to make operations easier. I'd much rather spin up my three new servers as non-voters and watch them to be healthy for a while before I put them into voting then to have to yolo like we do today. There are good applications for read scaling, too, with systems like Consul which allows stale reads.

> + - Is there a good way to break this up?

It seems like much of this could be done outside of `raft.go`. Ideally, we could have the peer state machine and "eligible to vote" algorithm in completely separate classes and testable outside of an integrated configuration. Similar for the log / snapshot configuration searcher (there will be quite a few cases to test there). 

> + - What do we need to worry about in terms of backwards compatibility?

Put some notes up above. I think we can fairly easily support `AddPeer()` during this transition, and it might be useful to hang on to the `PeerStore` to make migration easier. If something is going to be a huge effort to support we could probably break an interface, but it would be good to avoid it if we can, especially during this transition period.

We might need to be creative to support a zero-downtime rolling upgrade if we have a new type of config message. This would be ideal if we can figure out a way to do it.

> + - What implication will this have on current tests?

If we mirrored the committed configuration into the peer store (if one is supplied) we could probably get the existing `raft_test.go` to pass with the new code. This might give us a good baseline before we change all the tests to use whatever the new "who are your peers" interface is, then we could rip out the mirroring.